### ISSUE TITLE
Sending enum-arguments as variables (with Relay) mostly results in VariableTypeMismatch

### ISSUE THREAD

**Issue by fibbers** [graphql-java/graphql-java#203]
This issue is a bit related to the question and answer on http://stackoverflow.com/questions/35946313/how-to-set-a-relay-variable-as-an-enum-value/35987582#35987582, however my issue is specifically server-side.

I have these relevant parts of the schema for `Users`:

``` graphql
Users(orderBy: [UserOrdering]): UserConnection

input UserOrdering {
  sort: UserSortField!
  direction: SortDirection!
}

enum UserSortField {
  FIRSTNAME
  LASTNAME
  FULLNAME
}

enum SortDirection {
  ASC
  DESC
}
```

When sending the query with `orderBy` specified as a variable:

``` graphql
query($orderBy: [UserOrdering]!) {
  viewer {
    ...on LoggedInViewer {
      Users(orderBy:$orderBy) {
        edges {
          node {
            fullName
          }
        }
      }
    }
  }
}

variables:
{
  "orderBy": [{"sort": "FULLNAME", "direction": "ASC"}]
}
```

Most of the times I get back: `Validation error of type VariableTypeMismatch: Variable type doesn't match`.

When sending `orderBy` hardcoded, the query always works:

``` graphql
query {
  viewer {
    ...on LoggedInViewer {
      Users(orderBy: [{sort:FULLNAME, direction:ASC}]) {
        edges {
          node {
            fullName
          }
        }
      }
    }
  }
}
```

---

Now, the strange thing is that _sometimes_ querying with `orderBy` as a variable _works_ (when sending the query with GraphiQL a couple of times). I suspect it has something to do with the order of execution on the server-side, but I'm not sure what to look for exactly.

Can you help me making this query work consistently?

NB: I'm using version `2016-09-02T14-27-54` of `graphql-java`

**Comment by fibbers** [comment]
After debugging this a bit futher, I now see when the error is thrown.

Some debugging notes about the variables when evaluating [`VariableTypesMatchRule.checkVariable`](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/validation/rules/VariableTypesMatchRule.java#L37-L47):

<ul>
<li> this results in the error:

```
variableReference: VariableReference{name='orderBy'}
variableDefinition: VariableDefinition{name='orderBy', type=NonNullType{type=ListType{type=TypeName{name='UserOrdering'}}}, defaultValue=null}
variableType: GraphQLNonNull{wrappedType=graphql.schema.GraphQLList@259b9b2e}
  wrappedType: graphql.schema.GraphQLList@259b9b2e
    wrappedType: graphql.schema.GraphQLInputObjectType@259b9b2e
      name: “UserOrdering”
inputType: graphql.schema.GraphQLList@37d2f318
  wrappedType: graphql.schema.GraphQLInputObjectType@37d2f318
    name: “UserOrdering”
```
</li>

<li>this results in a succesful response:

```
variableReference: VariableReference{name='orderBy'}
variableDefinition: VariableDefinition{name='orderBy', type=NonNullType{type=ListType{type=TypeName{name='UserOrdering'}}}, defaultValue=null}
variableType: GraphQLNonNull{wrappedType=graphql.schema.GraphQLList@5cb3019a}
  wrappedType: graphql.schema.GraphQLList@5cb3019a
    wrappedType: graphql.schema.GraphQLInputObjectType@5cb3019a
      name: “UserOrdering”
inputType: graphql.schema.GraphQLList@5cb3019a
  wrappedType: graphql.schema.GraphQLInputObjectType@5cb3019a
    name: “UserOrdering”
```
</li>

Now, let's find out why `variableType.wrappedType` sometimes refers to a different object than `inputType` (which causes [`VariablesTypesMatcher`](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/validation/rules/VariablesTypesMatcher.java#L38) to error).

**Comment by okorz001** [comment]
I wonder if the issue is related to input objects? I am using enums heavily in my schema and have no issues. However, I am not using input objects at all.

**Comment by iatzmon** [comment]
I'm having the same problem using Apollo as my client

**Comment by fibbers** [comment]
I'm using Play Framework as my server. Seeing that this error is almost consistently reproducable (_sometimes_ all goes well though and only after trying a few times **on Dev mode***), I'm thinking maybe something weird is going on related to classloaders. Just a hunch though :-)

\* when sending a query, each time dev-mode seems to restart and reload classes, which is why I'm thinking it is classloader related. I haven't seen a succesful attempt on production, though.

**Comment by TomerSabags** [comment]
I had the same issue and managed to solve it, you might have the same problem - 
when creating the schema I used a method for creating each enum - so each time (for every query) this enum was created again and that's why the "==" check failed.
To solve it I created a map (while initializing the schema generator class) of every enum type in my schema and I defined the schema enums directly from this map, in that way the queries are all using exactly the same enum instance and the "==" check passes.
Hopefully this helps

**Comment by dminkovsky** [comment]
Ahhh thank you @TomerSabags. This makes sense... so we can add `VariableTypeMismatch` to the list of rules that depend on `==` for validation (#85).

**Comment by fibbers** [comment]
You pointed me in the right direction @TomerSabags, thanks!

I already created my schema as a Singleton instance to prevent re-creating the schema on each query. However, my schema had multiple [connections](https://facebook.github.io/relay/docs/graphql-connections.html) pointing to the same type, but using a different db-query (e.g. a User-connection from a Department-type, or a User-connection from a Project-type). Each User-connection accepts the same arguments, such as `UserOrdering`. This meant that if I had 3 User-connections in my schema, I also created 3 instances to handle the `UserOrdering` argument.

After caching and reusing the `GraphQLInputObjectType` for this `UserOrdering` argument when constructing the schema, the `==` checks now succeed and thus, sending enum-arguments as variables works!
